nixos/modules/services/security/vault.nix

```diff
@@ -0,0 +1,196 @@
+
+      ${if cfg.listener.tls_min_version != "tls12" then "tls_min_version = \"${cfg.listener.tls_min_version}\""  else ""} 
+
+      ${if cfg.listener.tls_cipher_suites != "" then "tls_cipher_suites = \"${cfg.listener.tls_cipher_suites}\"" else "" } 
+
+      ${if cfg.listener.tls_prefer_server_cipher_suites != "false" then "tls_prefer_server_cipher_suites = \"${cfg.listener.tls_prefer_server_cipher_suites}\"" else "" } 
+
+      ${if cfg.listener.tls_require_and_verify_client_cert != "false" then "tls_require_and_verify_client_cert= \"${cfg.listener.tls_require_and_verify_client_cert}\"" else "" } 
+
+    }
+
+    storage "${cfg.storage.backend}" {
+      ${cfg.storage.extraConfig}
+    }
+
+    ${if cfg.telemetry.extraConfig != "" then "
+      telemetry { 
+        ${if cfg.telemetry.disable_hostname then "disable_hostname = \"true\"" else ""}
+        ${cfg.telemetry.extraConfig}
+      }" else ""}
+
+  '';
+
+in
+{
+  options = {
+
+    services.vault = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enables the vault daemon.
+        '';
+      };
+
+      listener = {
+
+        address = mkOption {
+          type = types.str;
+          default = "127.0.0.1:8200";
+          description = ''
+            The name of the ip interface to listen to.
+          '';
+        };
+
+        cluster_address = mkOption {
+          type = types.str;
+          default = "";
+          description = ''
+            The name of the address to bind to for cluster server-to-server requests.
+          '';
+        };
+
+        tls_cert_file = mkOption {
+          type = types.str;
+          default = "";
+          description = ''
+            The name of the crt file for the ssl certificate.
+          '';
+        };
+        
+        tls_key_file = mkOption {
+          type = types.str;
+          default = "";
+          description = ''
+            The name of the key file for the ssl certificate.
+          '';
+        };
+
+        tls_disable = mkOption {
+          type = types.str;
+          default = "false";
+          description = ''
+            Specifies if TLS will be disabled. Vault assumes TLS by default, so you must explicitly disable TLS to opt-in to insecure communication.
+          '';
+        };
+        
+        tls_min_version = mkOption {
+          type = types.str;
+          default = "tls12";
+          description = ''
+             The minimum supported version of TLS. Accepted values are "tls10", "tls11" or "tls12".
+          '';
+        };
+
+        tls_cipher_suites = mkOption {
+          type = types.str;
+          default = "";
+          description = ''
+             The list of supported ciphersuites as a comma-separated-list.
+          '';
+        };
+
+        tls_prefer_server_cipher_suites = mkOption {
+          type = types.str;
```

Comment:
This could be something like `types.nullOr (types.enum [ "tls10" "tls11" "tls12" ])`.